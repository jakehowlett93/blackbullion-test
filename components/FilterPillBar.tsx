import { View, Pressable, StyleSheet, Text } from "react-native";

import { PathwayFilters } from "@/types/types";
import { capitalizeFirstLetter } from "@/utils/utils";

interface FilterPillBarProps {
  filters: PathwayFilters;
  handleFilterPress: (filter: keyof PathwayFilters) => void;
}

export const FilterPillBar = ({
  filters,
  handleFilterPress,
}: FilterPillBarProps) => {
  return (
    <View style={styles.filterContainer}>
      {Object.entries(filters).map(([filter, value]) => {
        const backgroundColor = value ? "#AD98F1" : "grey";
        return (
          <Pressable
            style={[styles.filterPills, { backgroundColor: backgroundColor }]}
            key={filter}
            onPress={() => handleFilterPress(filter as keyof PathwayFilters)}
          >
            <Text>{capitalizeFirstLetter(filter)}</Text>
          </Pressable>
        );
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  filterContainer: {
    marginHorizontal: 15,
    marginVertical: 5,
  },
  filterPills: {
    color: "white",
    borderRadius: 12,
    opacity: 0.7,
    alignSelf: "flex-start",
    paddingHorizontal: 10,
    paddingVertical: 5,
  },
});
