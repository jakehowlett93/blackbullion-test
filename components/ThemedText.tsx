import { Text, StyleSheet, TextProps } from 'react-native';

export type ThemedTextProps = TextProps & {
  type?: 'default' | 'title' | 'defaultSemiBold' | 'subtitle' | 'link' | 'overlay';
  color?: string;
};

export function ThemedText({
  style,
  color,
  type = 'default',
  ...rest
}: ThemedTextProps) {

return (
    <Text
      style={[
        { color },
        type === 'default' ? styles.default : undefined,
        type === 'title' ? styles.title : undefined,
        type === 'link' ? styles.link : undefined,
        type === 'overlay' ? styles.overlay : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  default: {
    fontSize: 16,
    lineHeight: 24,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    lineHeight: 32,
    color: '#663C97'
  },
  link: {
    lineHeight: 30,
    fontSize: 16,
    color: '#663C97'
  },
  overlay: {
    fontSize: 14,
    color: 'black',
  }
});
