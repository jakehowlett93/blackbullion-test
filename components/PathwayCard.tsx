import {
  View,
  StyleSheet,
  ImageBackground,
  Linking,
  Pressable,
} from "react-native";

import { Pathway } from "@/types/types";
import { ThemedText } from "./ThemedText";
import { useTheme } from "@react-navigation/native";
import { capitalizeFirstLetter } from "@/utils/utils";

interface PathwayCardProps {
  pathway: Pathway;
}

export const PathwayCard = ({
  pathway: { title, url, intro, duration, image, type },
}: PathwayCardProps) => {
  const { colors } = useTheme();

  return (
    <Pressable
      style={[styles.container, { backgroundColor: colors.card }]}
      onPress={() => Linking.openURL(url)}
      role="button"
      aria-label={`Navigate to ${title}`}
    >
      <View>
        <ImageBackground source={{ uri: image }} style={styles.image}>
          <View style={styles.pill}>
            <ThemedText type="overlay">{`${capitalizeFirstLetter(type)} - ${duration}`}</ThemedText>
          </View>
        </ImageBackground>
      </View>
      <View style={styles.content}>
        <ThemedText type="title">{title}</ThemedText>
        <ThemedText numberOfLines={3}>{intro}</ThemedText>
      </View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  container: {
    borderRadius: 12,
    height: 320,
    width: 300,
  },
  content: {
    margin: 8,
  },
  image: {
    height: 170,
    width: "100%",
    justifyContent: "flex-end",
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
    overflow: "hidden",
  },
  pill: {
    alignSelf: "flex-start",
    backgroundColor: "white",
    paddingHorizontal: 10,
    paddingVertical: 2,
    margin: 4,
    borderRadius: 8,
    justifyContent: "center",
    opacity: 0.7,
  },
});
