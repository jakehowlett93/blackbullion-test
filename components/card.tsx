import { View, StyleSheet, Image, Linking } from 'react-native';

import { Pathway } from '@/types/types';
import { ThemedText } from './ThemedText';

interface PathwayCardProps {
  pathway: Pathway
}

export const PathwayCard = ({pathway}: PathwayCardProps) => {
  const {title, url, intro, duration, image, type} = pathway;

  return (
    <View style={styles.container}>
      <Image source={{uri: image}} style={{height: 60, width: 200, resizeMode: 'contain'}} />
      <View style={{flexDirection: 'row'}}>
        <ThemedText style={{color: 'grey', fontSize: 8}}>{type}</ThemedText>
        <ThemedText>{duration}</ThemedText>
      </View>
      <ThemedText type='title'>{title}</ThemedText>
      <ThemedText>{intro}</ThemedText>
      <ThemedText type='link' onPress={() => Linking.openURL(url)}>View pathway</ThemedText>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    flexDirection: 'column',
    height: 300,
    width: 200,
    margin: 30
  } 
});
