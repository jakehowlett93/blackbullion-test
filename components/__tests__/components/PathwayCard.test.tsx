import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { PathwayCard } from "@/components/PathwayCard";
import { Linking } from "react-native";

const mockPathway = {
  id: 24,
  title: "Starting University",
  internal_title: "Starting University",
  url: "https://www.blackbullion.com/pathways/starting-university-uk",
  intro:
    "Learn what funding is available when you start uni and how you can get the best out of it.",
  duration: "11 min",
  image: "https://prodcontent.blackbullion.com/images/pathways/24/thumb",
  type: "pathway",
  has_summative_assessment: true,
};

jest.mock("react-native/Libraries/Linking/Linking", () => ({
  openURL: jest.fn(),
}));

describe("PathwayCard", () => {
  it("renders without errors", () => {
    const { getByText } = render(<PathwayCard pathway={mockPathway} />);

    expect(getByText("Starting University")).toBeTruthy();
    expect(
      getByText(
        "Learn what funding is available when you start uni and how you can get the best out of it.",
      ),
    ).toBeTruthy();
    expect(getByText("Pathway - 11 min")).toBeTruthy();
  });

  it("opens pathway URL on press", () => {
    const { getByRole } = render(<PathwayCard pathway={mockPathway} />);

    fireEvent.press(getByRole("button"));

    expect(Linking.openURL).toHaveBeenCalledWith(
      "https://www.blackbullion.com/pathways/starting-university-uk",
    );
  });

  it("matches snapshot", () => {
    const { toJSON } = render(<PathwayCard pathway={mockPathway} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
