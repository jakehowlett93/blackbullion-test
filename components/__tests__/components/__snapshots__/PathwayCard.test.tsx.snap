// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathwayCard matches snapshot 1`] = `
<View
  accessibilityLabel="Navigate to Starting University"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
  style={
    [
      {
        "borderRadius": 12,
        "height": 320,
        "width": 300,
      },
      {
        "backgroundColor": "rgb(255, 255, 255)",
      },
    ]
  }
>
  <View>
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        {
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "height": 170,
          "justifyContent": "flex-end",
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <Image
        source={
          {
            "uri": "https://prodcontent.blackbullion.com/images/pathways/24/thumb",
          }
        }
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "height": 170,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "alignSelf": "flex-start",
            "backgroundColor": "white",
            "borderRadius": 8,
            "justifyContent": "center",
            "margin": 4,
            "opacity": 0.7,
            "paddingHorizontal": 10,
            "paddingVertical": 2,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": undefined,
              },
              undefined,
              undefined,
              undefined,
              {
                "color": "black",
                "fontSize": 14,
              },
              undefined,
            ]
          }
        >
          Pathway - 11 min
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "margin": 8,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": undefined,
          },
          undefined,
          {
            "color": "#663C97",
            "fontSize": 20,
            "fontWeight": "bold",
            "lineHeight": 32,
          },
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      Starting University
    </Text>
    <Text
      numberOfLines={3}
      style={
        [
          {
            "color": undefined,
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      Learn what funding is available when you start uni and how you can get the best out of it.
    </Text>
  </View>
</View>
`;
